### PR TITLE
Run the Initial Setup TUI on all usable consoles (#1438046)

### PR DESCRIPTION
**Pull request by M4rtinK**
This change attempts to address two issues with the Initial Setup TUI:

1) starting the TUI on a different console than the one a user
   connected to
2) starting the Initial Setup TUI on a console that is unusable.

Number 1) might generally not be a big issue on x86 PCs, where
Initial Setup is only used sporadically for simple post installation
configuration and where most users are simply using the graphical
console.

Things are much more serious for arm based single board
computers, where Initial Setup TUI more or less takes over the role of
an interactive installer, as the main installation method is image
based and the Initial Setup TUI running in reconfig mode provides
the only interactive configuration options.

Further users of arm devices make much more use of serial consoles, so
the x86 platform expectation that the user is mostly likely at the
graphical console does not hold.

Issue number 2) also has increased impact on arm devices, as arm boards
tend to have various esoteric serial consoles, some of which might not
be useable in an interactive mode. If such console is resolved as the
default one then the Initial Setup TUI might fail to open the console
and crash at startup.

The solution - run IS TUI on all useable TTYs

This should solve issue number 1) by providing access the Initial Setup
TUI regardles to what console (graphical, serial, etc.) the user is
using. It should also solve issue number 2) as consoles that can't be
opened or that fail during use are skipped.

The implementation - stdin and stdout redirection and multiplexing

The solution to issues 1) & 2) is achieved by redirecting the Initial
Setup TUI stdout and stdin to multiplexing code, which uses select() to:

- watch for any output and sends it to all useable consoles at once
- watch for input on any useable console and sends it to the
  Initial Setup stdin.

This efectively means that the Initial Setup TUI should be running on
all useable consoles and the user should be able to use any of the
consoles to interact with it.

### CODE REVIEW

**Comment by M4rtinK** [comment]
An scratch build for easy testing:

https://koji.fedoraproject.org/koji/taskinfo?taskID=19509426

Note: Due to an unrelated Anaconda bug you need to run Initial Setup in reconfig mode:

touch /.unconfigured
then reboot

Otherwise only a single spoke will be displayed on the Initial Setup TUI hub, triggering the Aanconda side bug.

**Comment by M4rtinK** [comment]
Fix the single-spoke-on-TUI-hub issue: https://github.com/rhinstaller/anaconda/pull/1060

File: initial_setup/tui/tui.py [hunk 1]
```diff
@@ -11,6 +17,116 @@
 QUIT_MESSAGE = N_("Are you sure you want to quit the configuration process?\n"
                   "You might end up with unusable system if you do.")
 
+class MultipleTTYHandler(object):
+    """Run the Initial Setup TUI on all useable consoles.
```
Comment (vpodzime) [inline initial_setup/tui/tui.py]: *usable*

File: initial_setup/tui/tui.py [hunk 2]
```diff
@@ -11,6 +17,116 @@
...
+
+    This is done by redirecting the Initial Setup stdout to all
+    useable consoles and then redirecting any input back to
```
Comment (vpodzime) [inline initial_setup/tui/tui.py]: same here and in many other places below

File: initial_setup/tui/tui.py [hunk 3]
```diff
@@ -11,6 +17,116 @@
...
+
+    def shutdown(self):
+        """Tell the multi TTY handler to shutdown."""
+        self._shutdown = True
+
+    def _open_all_consoles(self):
+        """Open all consoles suitable for running the Initial Setup TUI."""
+        console_write_fos = []
+        console_read_fds = {}
+        console_paths = [os.path.join("/dev", c) for c in list_useable_consoles_for_tui()]
```
Comment (vpodzime) [inline initial_setup/tui/tui.py]: this could be a generator

File: initial_setup/tui/tui.py [hunk 4]
```diff
@@ -11,6 +17,116 @@
...
+        for console_path in console_paths:
+            try:
+                write_fo = open(console_path, "w")
+                read_fo = open(console_path, "r")
+                console_write_fos.append(write_fo)
+                read_fd = read_fo.fileno()
+                # the console stdin file descriptors need to be non-blocking
+                os.set_blocking(read_fd, False)
+                console_read_fds[read_fd] = read_fo
+                # As we survived till now the console might be useable
```
Comment (vpodzime) [inline initial_setup/tui/tui.py]: *If we...*

File: initial_setup/tui/tui.py [hunk 5]
```diff
@@ -11,6 +17,116 @@
...
+        log.info("multi TTY handler starting")
+        while True:
+            # Watch the consoles and IS TUI stdout for data and
+            # react accordingly.
+            # The select also triggers every second (the 1.0 parameter),
+            # so that the infinite loop can be promptly interrupted once
+            # the multi TTY handler is told to shutdown.
+            rlist, _wlist, _xlist = select.select(fds, [], [], 1.0)
+            if self._shutdown:
+                log.info("multi TTY handled shutting down")
```
Comment (vpodzime) [inline initial_setup/tui/tui.py]: *handler*

File: initial_setup/tui/tui.py [hunk 6]
```diff
@@ -11,6 +17,116 @@
...
+                    # as otherwise reading from it would (predictably) result in
+                    # the readline() function blocking once it runs out of data.
+                    os.set_blocking(fd, False)
+                    # The IS TUI wants to write something,
+                    # wrote that to all console fds.
+                    lines = []
+                    line = self._tui_stdout.readline()
+                    while line != "":
+                        lines.append(line)
+                        line = self._tui_stdout.readline()
```
Comment (vpodzime) [inline initial_setup/tui/tui.py]: ``lines = self._tui_stdout.readlines()``

File: initial_setup/tui/tui.py [hunk 7]
```diff
@@ -11,6 +17,116 @@
...
+                    os.set_blocking(fd, True)
+                    lines.append("\n")  # seems to get lost somewhere on the way
+                    for console_fo in self._console_write_fos:
+                        for one_line in lines:
+                            try:
+                                console_fo.write(one_line)
+                            except OSError:
+                                log.exception("failed to write %s to console %s", line, console_fo)
+                else:
+                    # Someone typed something to a consoles and hit enter,
```
Comment (vpodzime) [inline initial_setup/tui/tui.py]: *console*

File: systemd/initial-setup.service [hunk 8]
```diff
@@ -13,6 +13,7 @@ Type=oneshot
 TimeoutSec=0
 StandardInput=tty
 StandardOutput=tty
+LimitNOFILE=9001
```
Comment (vpodzime) [inline systemd/initial-setup.service]: Why ``9001``? I think this line needs an explanatory comment

Comment (M4rtinK) [inline systemd/initial-setup.service]: Oops - that's actually a debuging leftover and should not be needed - I'll remove it. :)

And why 9001 ? Well, [that's obvious](https://www.youtube.com/watch?v=SiMHTK15Pik). ;-)

File: initial_setup/tui/tui.py (hunk 1 above)
Comment (M4rtinK) [inline initial_setup/tui/tui.py]: Good point, that sounds, better - changing this to usable.

File: initial_setup/tui/tui.py (hunk 2 above)
Comment (M4rtinK) [inline initial_setup/tui/tui.py]: ...in all places. :)

File: initial_setup/tui/tui.py (hunk 4 above)
Comment (M4rtinK) [inline initial_setup/tui/tui.py]: OK.

File: initial_setup/tui/tui.py (hunk 3 above)
Comment (M4rtinK) [inline initial_setup/tui/tui.py]: Sure, why not. :)

File: initial_setup/tui/tui.py (hunk 5 above)
Comment (M4rtinK) [inline initial_setup/tui/tui.py]: Fixed.

File: initial_setup/tui/tui.py (hunk 7 above)
Comment (M4rtinK) [inline initial_setup/tui/tui.py]: Fixed. :)

File: initial_setup/tui/tui.py (hunk 6 above)
Comment (M4rtinK) [inline initial_setup/tui/tui.py]: Good point & seems to work fine. :)

**Pull request merged**